### ISSUE TITLE
More docs/support for JavaScript

### ISSUE THREAD

**Issue by missmatsuko** [doublesecretagency/craft-smartmap#30]
I'm creating an interactive/searchable map. It works fine when I build it out in the template with Twig, but I'd really like to use JavaScript so it doesn't require page reload to get new results.

The JavaScript examples in the docs are a bit sparse and scattered. I did try to use them, but haven't had much luck.

- From the [Manipulating the Map in JavaScript](https://www.doublesecretagency.com/plugins/smart-map/docs/manipulating-the-map-in-javascript) page, I tried adding/removing map markers and refreshing the map. It didn't seem to change the visible map markers, but I did see the items listed change when I ran `smartMap.listMarkers();`
- I can't get [Lookup via AJAX](https://www.doublesecretagency.com/plugins/smart-map/docs/front-end-address-lookup) to work at all. I even tried including jQuery and copy and pasting the example. I keep getting 404 errors for the endpoint even though I'm sending a CSRF token and a target. I also tried making the path relative to root (e.g. `/actions/smartMap/lookup` instead of `actions/smartMap/lookup`).
- I don't think there's a way to get information about the markers already loaded on the map. Although, if I could get Lookup via AJAX working, then I could render the map and keep track of the markers myself.
- Not sure if there's a way to get the [Visitor Geolocation](https://www.doublesecretagency.com/plugins/smart-map/docs/visitor-geolocation) with JavaScript, or if it requires a custom plugin.
- It would be nice to be able to use your [address formatting](https://www.doublesecretagency.com/plugins/smart-map/docs/automatically-format-an-entire-address) from JavaScript as well, as I'm pretty sure Google doesn't provide addresses formatted with line breaks.

**Comment by lindseydiloreto** [comment]
>The JavaScript examples in the code are a bit sparse and scattered.

I actually agree with that, I apologize. I have long-term plans to improve the JS docs, but it sadly won't be anytime soon. In the meantime, feel free to ping me with questions either here or on Slack. 🙂 

The good news is, the `smartMap` JS is a relatively thin layer on top of the entire Google Maps API! Anything you can do with the native Google JS, you should be able to do using the `smartMap.map` and `smartMap.marker` data.

1. It sounds like you were able to remove the marker object from the `smartMap.marker` array, but didn't actually remove the marker from the _map_. That's handled separately, using something like [this...](https://developers.google.com/maps/documentation/javascript/examples/marker-remove)
    ```javascript
    smartMap.marker['smartmap-mapcanvas-1.33.myAddressField'].setMap(null);
    ```

2. Are you using a different [`actionTrigger`](https://docs.craftcms.com/v3/config/config-settings.html#actiontrigger)? If not, I don't see an obvious reason why this wouldn't work. Ping me on Slack and we can troubleshoot it together.

3. The `smartMap.marker` array is a collection of markers, but not necessarily in relation to a specific map. `smartMap.map` is an array of all maps. Beyond that, you'd just use normal Google Maps JS methods to interact with those objects.

4. You could potentially retrieve the visitor geolocation data via a simple AJAX call. Create a template which only contains this...
    ```twig
    {{ craft.smartMap.visitor | json_encode }}
    ```
    ... then use your AJAX call to retrieve and decode it.

5. The address formatting logic is pretty rudimentary... feel free to [mimic it](https://github.com/doublesecretagency/craft-smartmap/blob/v3/src/models/Address.php#L125-L151) in your own JS function.

Hope that helps!

**Comment by missmatsuko** [comment]
Thanks, @lindseydiloreto 

Re 1 and 3: I'm wondering what is the use case for the marker functions in JavaScript since it doesn't manipulate the map directly and doesn't seem to provide useful info to use for manipulating the map? Maybe I'm missing something there.

Re 2: No, I didn't change the actionTrigger. I'll try to get in touch with you on Slack to troubleshoot further. Thanks!

**Comment by lindseydiloreto** [comment]
The Google models aren't tied directly to the DOM. You'd still need to use Google's API to actually _manipulate_ those objects (and in turn, the DOM).

You may be thinking in terms of React or Vue, where there's an invisible link between DOM elements and JS models. That doesn't really apply here... you need to take specific actions on a model in order for it to affect the DOM.

**Comment by missmatsuko** [comment]
@lindseydiloreto That's fine, I'm just not sure what the functions are useful for if they don't manipulate the map. I also think the docs could be clearer about what the functions do, especially since the name of the page is "Manipulating the map in JavaScript".

**Comment by lindseydiloreto** [comment]
Oh, you're talking about the "Additional Functions" section at the bottom? Specifically, are you having trouble with the `createMarker` and `deleteMarker` functions?

Those should work as expected (by changing the map). Can you show me your code calling those two functions?

**Comment by missmatsuko** [comment]
Hmm, actually trying it again now and `deleteMarker` is working as expected. 

I did get errors with `createMarker`, but I guess I need to pass in the map element's ID and the actual coordinates of the marker in the `options` parameter. It's a bit unintuitive that `createMarker` and `deleteMarker` require different information. 

I expected `createMarker` to query for the info it needs and add the marker to the appropriate map based on its `markerName` since that could be parsed from the format `<MAP ID>.<ELEMENT ID>.<FIELD HANDLE>`.

So I've tried adding the marker back in that I deleted, and refreshing the map, but I'm not seeing a new marker on the map. I think it should at least show a new marker with the Google map's default map pin?
```
// Start on page load, map has one marker on it

smartMap.listMarkers(); // ["smartmap-mapcanvas-1.1298.locationAddress"]

smartMap.deleteMarker("smartmap-mapcanvas-1.1298.locationAddress");
// Marker is visibly remove from map, as expected

smartMap.listMarkers(); // ["smartmap-mapcanvas-1.1298.locationAddress"]
// Expected this to be an empty array?

smartMap.createMarker("smartmap-mapcanvas-1.1298.locationAddress", {
    mapId: "smartmap-mapcanvas-1",
    position: {lat: 60.71917590, lng: -135.05363010},
});
// [smartmap-mapcanvas-1.1298.locationAddress] Marker rendered.
// No errors, no new map pin

smartMap.refreshMap("smartmap-mapcanvas-1");
// Still no errors, no new map pin
```

It's all kind of moot if I can get the AJAX lookup working, though. Then I can handle rendering the maps and markers with Google's API directly.

**Comment by lindseydiloreto** [comment]
To be honest, I'll probably do a major rewrite of the entire front-end JS in a future version. You may not want to get too hung up on the details of these particular functions.

Those functions were all intended as simple helper functions, feel free to re-declare or copy them as needed. They were written _very_ early on in Smart Map 2, and may get deprecated in Smart Map 4. 

Feel free to ping me on Slack, I'm happy to help work out the AJAX issue! 🙂

**Comment by JyroSnoeijen** [comment]
I also had the problem with the lookup action,
the url should be changed to 'actions/smart-map/lookup' (for craft3 it must be in snake-case),
and the csrfToken should be added to the ajax header.

In the lookupController requireAcceptsJson() is used,
But the data isnt handled like json.
Can the code for the LookupController be changed to:

````
public function actionIndex()
    {
        $this->requirePostRequest();
        $this->requireAcceptsJson();

        $request = Craft::$app->getRequest();
        $payload = Json::decode($request->getRawBody(), true);
        $target = (isset($payload['target']) ? $payload['target'] : null);
        $components = (isset($payload['components']) ? $payload['components'] : null);
        $components = $this->_explodeComponents($components);
        $response = SmartMap::$plugin->smartMap->lookup($target, $components);
        return $this->asJson($response);
    }
````
?

Working example code for using with jquery after the changes above:
````
var target = $('#street1').val() + ' ' + $('#zip').val() + ' ' + $('#city').val() + ' ' + $('#country').val();

var csrfTokenValue = "{{ craft.app.request.csrfToken }}";

$.ajax({
    headers: {'X-CSRF-TOKEN': csrfTokenValue}
    type: 'post',
    url: 'actions/smart-map/lookup/',
    dataType: 'json',
    contentType: 'application/json; charset=utf-8',
    data: JSON.stringify({
        target: target
    }),
    success: function(data){
        console.log(data['results'][0]['geometry']['location']);
    },
    failure: function(errMsg) {
    }
});
````

**Comment by missmatsuko** [comment]
Looks like that's the same problem I was having. I tried just changing the endpoint to snake-case and got a 400 response (Bad Request: Request must accept JSON in response) rather than 404.

**Comment by lindseydiloreto** [comment]
Great, thanks for the detailed report @JyroSnoeijen! (And thanks for confirming @missmatsuko!) I've just done the quick fix of updating the docs on our site.

https://www.doublesecretagency.com/plugins/smart-map/docs/front-end-address-lookup

I'll take a look into the suggested PHP changes. The _quickest_ change may simply be to remove the `requireAcceptsJson` line. But I'll do some testing before I officially change anything.

Will report back here with the final resolution. 👍

**Comment by lindseydiloreto** [comment]
This has been fixed in **v3.2.2**

I ended up simply removing the `requireAcceptsJson` requirement. You should now be able to perform the lookup exactly as described in the [docs](https://www.doublesecretagency.com/plugins/smart-map/docs/front-end-address-lookup) (which have been updated).

Thank you both for reporting this! 🙂
